Samsung AC Display and VirusDoctor toggles

rg2k opened the pull request:
1) Adds the capability to turn on/off the AC display. Useful if you sleep with the AC on.
2) Adds the capability to turn on/off the VirusDoctor feature.

VirusDoctor generates active hydrogen and oxygen ion, which forms Hydroperoxy radicals (HOO–). It reacts with virus and turns into harmless water vapor (H2O) Also the active hydrogen (H) combines with OH-radical and forms water vapor, neutralizing harmful OH-radical.

Tested on Samsung AC model: AR09FSSDAWKNFA with remote: DB63-03556X003.

rg2k committed:
Adding Display and Virus Doctor const

rg2k committed:
Adding code for Display and VirusDoctor

crankyoldgit commented on src/ir_Samsung.h:
```diff
@@ -99,6 +103,12 @@ class IRSamsungAc {
+  bool getDisplay();
+  void setVirusDoctor(const bool state);
+  bool getVirusDoctor();
+  
```
src/ir_Samsung.h:111:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]

crankyoldgit commented on src/ir_Samsung.cpp:
```diff
@@ -570,6 +570,30 @@ void IRSamsungAc::setPowerful(const bool on) {
 }
 
+bool IRSamsungAc::getDisplay(void) {
+  return remote_state[10];
```
This is wrong. It will always be `true` as it is always non-zero.
Use:
```
  return GETBIT8(remote_state[10], kSamsungAcDisplayOffset);
```

crankyoldgit commented on src/ir_Samsung.cpp:
```diff
@@ -570,6 +570,30 @@ void IRSamsungAc::setPowerful(const bool on) {
+  return remote_state[10];
+}
+
+void IRSamsungAc::setDisplay(const bool state) {
```
Change `state` to `on`.

crankyoldgit commented on src/ir_Samsung.cpp:
```diff
@@ -570,6 +570,30 @@ void IRSamsungAc::setPowerful(const bool on) {
+    remote_state[10] = kSamsungAcDisplaykOn;
+    } else {
+    remote_state[10] = kSamsungAcDisplaykOff;
+  }
```
Replace this with:
```
  setBit(&remote_state[10], kSamsungAcDisplayOffset, on);
```

crankyoldgit commented on src/ir_Samsung.h:
```diff
@@ -62,6 +62,10 @@ const uint8_t kSamsungAcPowerful10Size = 1;  // Mask 0b00000110
 const uint64_t kSamsungAcPowerSection = 0x1D20F00000000;
 
+const uint8_t kSamsungAcDisplaykOn = 0x71;
+const uint8_t kSamsungAcDisplaykOff = 0x61;
```
Replace with:
`const uint8_t kSamsungAcDisplayOffset = 4;  // 0b00010000`
as that is the only bit that changes between `0x71` & `0x61`

crankyoldgit commented on src/ir_Samsung.h:
```diff
@@ -62,6 +62,10 @@ const uint8_t kSamsungAcPowerful10Size = 1;  // Mask 0b00000110
 
+const uint8_t kSamsungAcDisplaykOn = 0x71;
+const uint8_t kSamsungAcDisplaykOff = 0x61;
+const uint8_t kSamsungAcVirusDoctorMask = 0x01;
```
replace with: `const uint8_t kSamsungAcIonOffset = 0;  // 0b00000001`

crankyoldgit commented on src/ir_Samsung.h:
```diff
@@ -99,6 +103,12 @@ class IRSamsungAc {
+  void setDisplay(const bool state);
+  bool getDisplay();
+  void setVirusDoctor(const bool state);
+  bool getVirusDoctor();
```
Please use `setIon(const bool on);` & `getIon(void);` instead. This makes it similar to other A/C Units.

crankyoldgit commented on src/ir_Samsung.cpp:
```diff
@@ -570,6 +570,30 @@ void IRSamsungAc::setPowerful(const bool on) {
+  return remote_state[11] & kSamsungAcVirusDoctorMask;
+}
+
+void IRSamsungAc::setVirusDoctor(const bool state) {
```
To be more consistent with other A/C units please use: `void IRSamsungAc::setIon(const bool on) {`

crankyoldgit commented on src/ir_Samsung.cpp:
```diff
@@ -570,6 +570,30 @@ void IRSamsungAc::setPowerful(const bool on) {
+    remote_state[11] |= kSamsungAcVirusDoctorMask;
+    } else {
+    remote_state[11] &= ~kSamsungAcVirusDoctorMask;
+  }
```
This can be replaced by:
```
  setBit(&remote_state[11], kSamsungAcIonOffset, on);
```

crankyoldgit commented on src/ir_Samsung.cpp:
```diff
@@ -570,6 +570,30 @@ void IRSamsungAc::setPowerful(const bool on) {
+}
+
+bool IRSamsungAc::getVirusDoctor(void) {
+  return remote_state[11] & kSamsungAcVirusDoctorMask;
```
Use this instead:
```
return GETBIT8(remote_state[11], kSamsungAcIonOffset);
```

crankyoldgit reviewed (changes requested):
Thanks a lot for this addition!
Mostly okay. Just some library-isms & code style/re-use things to correct.
Sorry for the nitpicks! :-)

rg2k commented:
> Thanks a lot for this addition!
> Mostly okay. Just some library-isms & code style/re-use things to correct.
> Sorry for the nitpicks! :-)

Thanks for the nitpicks and apologies for being such a noob. Will get it sorted.

rg2k committed:
Fixed issues as per Crankoldgit's request

rg2k committed:
Removed commented out code

rg2k committed:
Fixed spelling mistake

crankyoldgit committed:
Fix linter issues

src/ir_Samsung.h:67:  At least two spaces is best between code and comments  [whitespace/comments] [2]
src/ir_Samsung.h:68:  At least two spaces is best between code and comments  [whitespace/comments] [2]
src/ir_Samsung.h:112:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]

crankyoldgit reviewed (approved):
Thanks.
Looks fine now. Just fixing some linter issues and we are good to go!

crankyoldgit commented:
Xref: #1045

crankyoldgit commented:
The changes mentioned above have been included in the latest [IRremoteESP8266 v2.7.4](https://github.com/crankyoldgit/IRremoteESP8266/releases/tag/v2.7.4) release.
